### PR TITLE
[GOBBLIN-2030] Integrate AutoTroubleshooter with Temporal

### PR DESCRIPTION
**Pull request by homatthew**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-XXX

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/activity/impl/ProcessWorkUnitImpl.java [hunk 1]
```diff
@@ -66,13 +70,36 @@ public class ProcessWorkUnitImpl implements ProcessWorkUnit {
 
   @Override
   public int processWorkUnit(WorkUnitClaimCheck wu) {
+    AutomaticTroubleshooter troubleshooter = null;
+    Optional<EventSubmitter> eventSubmitter = Optional.ofNullable(wu.getEventSubmitterContext()).map(EventSubmitterContext::create);
```
Comment (phet) [inline gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/activity/impl/ProcessWorkUnitImpl.java:74]: `ofNullable`?  wasn't there `@NonNull` annotation on that member?

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/activity/impl/ProcessWorkUnitImpl.java [hunk 2]
```diff
@@ -66,13 +70,36 @@ public class ProcessWorkUnitImpl implements ProcessWorkUnit {
...
     try (FileSystem fs = Help.loadFileSystemForce(wu)) {
       List<WorkUnit> workUnits = loadFlattenedWorkUnits(wu, fs);
       log.info("(M)WU [{}] - loaded; found {} workUnits", wu.getCorrelator(), workUnits.size());
       JobState jobState = Help.loadJobState(wu, fs);
-      return execute(workUnits, wu, jobState, fs);
+      log.info("Loaded jobState. Instantiating and starting troubleshooter...");
+      troubleshooter = AutomaticTroubleshooterFactory.createForJob(ConfigUtils.propertiesToConfig(jobState.getProperties()));
+      troubleshooter.start();
+      log.info("Finished instantiating troubleshooter. Now executing flow");
```
Comment (phet) [inline gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/activity/impl/ProcessWorkUnitImpl.java:82]: is there really value in `.info` level debugging?  won't exception logging let us know where something went wrong (following line 77's logging)?

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/activity/impl/ProcessWorkUnitImpl.java [hunk 3]
```diff
@@ -66,13 +70,36 @@ public class ProcessWorkUnitImpl implements ProcessWorkUnit {
...
+      return execute(workUnits, wu, jobState, fs, troubleshooter.getIssueRepository());
     } catch (IOException | InterruptedException e) {
       throw new RuntimeException(e);
+    } finally {
+      try {
+        if (troubleshooter == null) {
+          log.warn("No troubleshooter to report issues from automatic troubleshooter");
+        } else {
+          log.info("Refining issues and logging summary...");
```
Comment (phet) [inline gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/activity/impl/ProcessWorkUnitImpl.java:91]: let's consistently use a prefix like on line 77:
```
log.info("(M)WU [{}] - ... ", wu.getCorrelator(), ...);
```

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/activity/impl/ProcessWorkUnitImpl.java [hunk 4]
```diff
@@ -66,13 +70,36 @@ public class ProcessWorkUnitImpl implements ProcessWorkUnit {
...
+          troubleshooter.refineIssues();
+          troubleshooter.logIssueSummary();
+          if (eventSubmitter.isPresent()) {
```
Comment (phet) [inline gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/activity/impl/ProcessWorkUnitImpl.java:94]: AFAICT this shouldn't be `Optional` - see earlier comment

**Review by phet — COMMENTED** [review]
looks like a great start - nice work, matt!

is it working when you execute this, so that the "issues" show up in gaas?  do they also show up in the exec logs?

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/work/EagerFsDirBackedWorkUnitClaimCheckWorkload.java [hunk 5]
```diff
@@ -30,15 +34,17 @@
 @lombok.NoArgsConstructor // IMPORTANT: for jackson (de)serialization
 @lombok.ToString(callSuper = true)
 public class EagerFsDirBackedWorkUnitClaimCheckWorkload extends AbstractEagerFsDirBackedWorkload<WorkUnitClaimCheck> {
```
[the remark this answers is not in the document]
Comment (homatthew) [reply gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/work/EagerFsDirBackedWorkUnitClaimCheckWorkload.java:36]: The workunit claimcheck is the input the exec workflow. The eventsubmitter context must be passed in as an argument to the workflow somehow, and it seemed like a reasonable spot

**Review by phet — APPROVED** [review]
great job - glad to reactivate this!

**Review by Will-Lo — APPROVED** [review]

**Pull request merged**